### ISSUE TITLE
Can Pihole works with a proxy like polipo ?

### ISSUE THREAD

**Issue by SR-G** [pi-hole/pi-hole#480]
##### Expected Behaviour:

I would like to have the following chain : 

client (web) --> using polipo proxy on a local server --> polipo uses  DNS from pihole (on same server) --> google DNS
##### Actual Behaviour:

Everything individually used works fine.
- polipo works fine when used directly with something like google DNS servers (8.8.8.8)
- pihole works fine when used direcly on the client (through resolv.conf on the client)

But when i link polipo with pihole (putting the server's IP in the polipo's configuration, instead of the google DNS), nothing works anymore : i still see the DNS requests in the pihole admin panel (so polipo actually resolves its DNS names through pihole as expected), but nothing works on the client (like if polipo weren't getting its results from pihole).

I really don't know if it's a polipo error, pihole error, or something weird on my server (running archlinux, by the way).
##### Steps to reproduce this issue:

Install polipo in a docker container
Install the pihole docker container (alpine version)
Configure the DNS for polipo in its config to use the pihole DNS server (dnsNameServer = IP)
##

**Comment by dschaper** [comment]
Have you been clearing the cache on Polipo between DNS server changes?

**Comment by SR-G** [comment]
Thanks for the idea, I hadn't initially but i've just made another try 
- stop everything
- purge my polipo /cache/ folder
- purge my polipo container and recreate it

But same behavior. 

I still see my dns queries in my pihole docker container logs (queried by/through polipo), but nothing works, like if something was weird for polipo when getting back its resolved addresses. No logs / errors in polipo container (and polipo fail back to system or default DNS only if the first DNS is not available : if i shut down the pihole container, then, of course, everything works again).

My pihole container is runned as suggested with 53:53/tcp + 53:53/ucp ports.

**Comment by dschaper** [comment]
Does your Polipo container have an available shell that you could docker run a dig or nslookup command and see what exactly you re getting back for white and black DNS addresses?

**Comment by SR-G** [comment]
Yes i can of course enter the polipo container with a bash shell, any idea what to look for ?

I've put polipo in the highest log level (0xFF) but nothing shown about DNS resolution in those logs (excepted when pihole is stopped for ex.).

**Comment by dschaper** [comment]
Try running a few nslookup runs and see if you are getting back reasonable responses. Try to lookup Google and see if it's a public address and look up a domain that should be blocked and see if it's returning the address of the Pi-hole. It kind of sounds like all responses are coming back as Pi-holed?

**Comment by SR-G** [comment]
Yep, i just did that.
- putting back pihole as system DNS in resolv.conf
- nslookup on various never-used-before-sites with nslookups : i get the correct result in the container (and i see the request in the pihole admin panel)

So that container with pihole seems to be working as expected.

I've really the feeling that the only not-working thing is "pihole DNS under polipo" : in that situation, the requests are sent to pihole (i see them in the pihole admin panel), but for whatever reason, polipo doesn't get back or doesn't understand the response (and does not log anything about this). At this point it's a bit weird and it may be a pure polipo problem ... excepted that with another DNS server everything works fine for polipo too :/

**Comment by dschaper** [comment]
Okay, and can I ask what environment you are in? Raspberry or another platform, and what OS and version? And lastly, where did you get the containers and instructions, so I can see if I can duplicate the problem.

**Comment by SR-G** [comment]
Host : archlinux 64 bit LTS + docker version 1.10.3, build 20f81dd
(running on a regular i686 pc, not a raspberry pi)

pihole container taken from the official image (latest alpine version, re-pulled today).

polipo  installed in a docker container with this Dockerfile (that means, polipo (latest version at this time) is rebuilt from sources) :
```
FROM	alpine
VOLUME	["/cache", "/etc/polipo/"]
EXPOSE	8123
CMD	[ "/usr/local/bin/polipo", "-c", "/etc/polipo/config" ]
ADD	./etc /etc/ # just a default /etc/polipo/config file if not injected at runtime

ENV	POLIPO_VERSION	1.1.1

RUN	mkdir -p /etc/polipo /cache /opt/ && \
	apk --update add bash ca-certificates build-base texinfo git && \
	git clone https://github.com/jech/polipo.git /opt/polipo/ && \
	cd /opt/polipo/ && \
	git checkout polipo-$POLIPO_VERSION && \
	make EXTRA_DEFINES="-DCHUNK_SIZE=16384" && \
	make install
```

Nothing special about running polipo, just injecting host paths for cache and config :

```
docker run -p 8123:8123 -d -v HOST_CACHE_DIR:/cache -v HOST_POLIPO_CONFIG_DIR:/etc/polipo/ --name polipo polipo
```

and my polipo config is (filename "config") :
```
# This file only needs to list configuration variables that deviate
# from the default values.  See /usr/share/doc/polipo/examples/config.sample
# and "polipo -v" for variables you can tweak and further information.

# Note IPv6 is not enabled
proxyAddress = "0.0.0.0"
disableIndexing = false
disableServersList = false
diskCacheRoot = /cache/

chunkHighMark = 1073741824
objectHighMark = 16384

maxConnectionRequests = 1024
maxObjectsWhenIdle = 128
maxPipelineTrain = 32
maxWriteoutWhenIdle = 262144
serverSlots = 8
maxDiskEntries = 65536

dnsNameServer = 192.168.8.4

logLevel = 0xFF
```

(192.168.8.4 is my local server, running both polipo and pihole, so dnsNameServer is the line linking polipo with pihole (DNS resolves are made directly by polipo))

i've just made a try with a "ubuntu" base instead of an alpine one (for the polipo container) but it's the same.

**Comment by dschaper** [comment]
Thanks for the Dockerfiles, I'll spin up an arch environment and see if I can duplicate, may take a little bit of time but hopefully we'll find a resolution for the issue.

**Comment by dschaper** [comment]
We don't have an official Docker container for the Pi-hole, can you give me a link to the containers hub page and any instructions you used to install? And how did you install Arch, I can't seem to find an LTS image, it's a rolling release OS from my experience. I just want to make sure I'm building an environment that's as close to yours as possible. I'll probably be spinning up the latest tarball from Arch on a Vagrant install however, if I can.

**Comment by SR-G** [comment]
About the docker pihole image : thought it was an official one but indeed not. I took the published "diginc/pi-hole" (https://github.com/diginc/docker-pi-hole).

Started with 
```
docker run -p 80:80 -p 53:53/udp -p 53:53/tcp -d --cap-add=NET_ADMIN -e ServerIP="HOST_IP" -v HOST_PATH_TO_ERROR_LOG:/var/lib/nginx/logs/error.log -v HOST_PATH_TO_ACCESS_LOG:/var/lib/nginx/logs/access.log --name pihole diginc/pi-hole
```

About archlinux : it is indeed a rolling release OS, but on the installation step, you can install a LTS kernel (that's what the LTS is about) "pacman -S linux-lts" (https://www.archlinux.org/packages/core/x86_64/linux-lts/) (but at first thoughts, I don't think this issue may be host related).

**Comment by dschaper** [comment]
Thanks for the update, I'll take a look at what I can come up with. I didn't know about the LTS kernel for Arch, good to know! I'll keep you updated with what I find.

**Comment by dschaper** [comment]
Okay, I'm working on a build environment that we can use in the future as a test platform. It may take a bit to do that, so in the meantime I'll spin up a VirtualBox Arch and see if I can find where the problem lies. I'm thinking there may be some inter-container networking issues but I'm not entirely convinced yet...

**Comment by diginc** [comment]
Hi, I'm the docker dev.  I'd be glad to give this a shot too.  

This really screams docker issue too so if you want me to open an issue on my repo over at diginc/docker-pi-hole I'd be glad to too.

**Comment by diginc** [comment]
I can reproduce your problem using default docker networking.   Using --net=host on polipo fixed it while still using pihole.  This seems to point to docker-proxy / IPTABLEs issues, I'll see if I can come up with a root cause and or more elegant solution fix the problem another day.

**Comment by dschaper** [comment]
@diginc Hi there! Thanks for the (unofficial) Docker work. (Sorry, have to add in the unofficial part, but I don't mean it as a slight or an insult to you in the least.) I'm working through trying to learn Docker and all that fun stuff in my spare time. What do you think about breaking the container out into Composeable service containers? (Right now, the webserver and resolver are in the same container, and we're looking at options for different resolvers and different webservers. My little Docker knowledge wonders if we could either Compose or is the new preferred method to have the containers advertise their services? I have a lot more reading to do!)

**Comment by diginc** [comment]
@dschaper - no slight taken.  It seems my docker has grown a bit of an user base with 5k downloads, so I should be thanking you guys for taking on some of the support that my image has generated.  :bow: 

Having never ran pi-hole on RBP my self, I decided to not modify the code base any more than the bare minimum which meant making the docker container replace the RBP essentially.  I dug into the scripts a bit when I first wrote the docker and thought there'd have to be numerous have the shell+php code be docker ready and didn't want to fork.  

This 'Docker as a VM' multi service container is not a docker best practice while one process per VM is.  I've heard arguments for and against both and am more for the 1 process per container methodology.

Not forking has made updating with the latest code improvements the pi-hole team made super easy too.  I could have probably left the code unforked but automatically applied patches to 'dockerize' the necessary lines of code but opted to just leave things as untouched as possible.

The pi-hole image is the first "Docker as a VM" style (multi-service) image I've made and I'm surprised with how reliable it has been for me in my home use.  Since I got it working I hadn't really toyed with the idea of splitting it out into 2 compose containers again.

It'd be neat to refactor web/dns into separate containers but I question if running compose's yaml file definitions is easier or harder for new-ish docker users than just one `docker run` command.  Technically speaking it can definitely be done though, I'm still on the fence about if I would encourage it for an official pi-hole image or not though.

**Comment by diginc** [comment]
@SR-G I found a way to do polipo's DNS without changing the `--net` type of the container.  It is not too pretty but its a lot safer than a container with net=host

I looked up the internal docker IP of my pihole container and passed that into the polipo container for DNS, both to the polipo config and the docker --dns argument worked by them selves.  Here is my --dns cmd while I was testing:

`docker run -ti --rm -p 8080:8123 --name polipo -v `pwd`/polipo.conf:/etc/polipo/config --link lanhelpers_pihole_1 --dns 172.17.0.7 clue/polipo`

The link is probably not necessary if you have the less secure docker daemeon icc=true setting as described here : https://docs.docker.com/engine/userguide/networking/default_network/container-communication/#communication-between-containers

I don't by default in ubuntu but I left the link in there just to be extra explicit. 

Like I said before I'm pretty sure the original problem is something with docker's iptables rules and how outgoing traffic is routed from containers to forwards of other containers.  Bypassing the rules and going directly to the other docker container through the docker0 subnet IP and `--net host` to avoid NAT all together are one of many solutions here.  

`--net=pihole` (merge polipo's docker networking with pihole's) and using 'localhost/127.0.0.1' for the DNS would be another decent solution (I did not test but just thought of) if you wanted to merge the 2 container's port forward, you wouldn't need to know the IP of the pihole container then, just the name.
